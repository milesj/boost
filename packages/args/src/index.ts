/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import format from './format';
import parse from './parse';
import parseInContext from './parseInContext';
import ArgsError from './ArgsError';
import type { ArgsErrorCode } from './ArgsError';
import ParseError from './ParseError';
import ValidationError from './ValidationError';

export * from './constants';
export * from './types';

export { format, parse, parseInContext, ArgsError, ParseError, ValidationError };
export type { ArgsErrorCode };
