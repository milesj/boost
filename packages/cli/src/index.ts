/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './CLIError';
export * from './Command';
export * from './constants';
export { Config } from './decorators/Config';
export * from './helpers';
export * as Arg from './metadata/args';
export * from './middleware';
export * from './Program';
export * from './types';
