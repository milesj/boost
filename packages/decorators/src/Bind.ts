import isMethod from './helpers/isMethod';

export default function Bind(): MethodDecorator {
  return (target, property, descriptor) => {
    if (
      !isMethod(target, property, descriptor) ||
      !('value' in descriptor && typeof descriptor.value === 'function')
    ) {
      throw new TypeError(`\`@Bind\` may only be applied to class methods.`);
    }

    const func = descriptor.value;

    return {
      configurable: true,
      get(this: Function) {
        const bound = func.bind(this);

        // Only cache the bound function when in the deepest sub-class,
        // otherwise any `super` calls will overwrite each other.
        if (target.constructor.name === this.constructor.name) {
          Object.defineProperty(this, property, {
            configurable: true,
            value: bound,
            writable: true,
          });
        }

        return bound;
      },
    };
  };
}
