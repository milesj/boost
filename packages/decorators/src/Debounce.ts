import isMethod from './helpers/isMethod';

export default function Debounce(delay: number): MethodDecorator {
	return (target, property, descriptor) => {
		if (__DEV__ && (
				!isMethod(target, property, descriptor) ||
				!('value' in descriptor && typeof descriptor.value === 'function')
			)) {
				throw new TypeError(`\`@Debounce\` may only be applied to class methods.`);
			}

		// We must use a map as all class instances would share the
		// same timer value otherwise.
		const timers = new WeakMap<Function, NodeJS.Timeout>();

		// Overwrite the value function with a new debounced function
		const func = descriptor.value;

		// @ts-expect-error
		descriptor.value = function debounce(this: Function, ...args: unknown[]) {
			const timer = timers.get(this);

			if (timer) {
				clearTimeout(timer);
				timers.delete(this);
			}

			timers.set(
				this,
				setTimeout(() => {
					(func as unknown as Function).apply(this, args);
				}, delay),
			);
		};
	};
}
