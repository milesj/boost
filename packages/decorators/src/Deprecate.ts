/* eslint-disable no-console */

import isClass from './helpers/isClass';
import isMethod from './helpers/isMethod';
import isProperty from './helpers/isProperty';
import isParam from './helpers/isParam';

export default function Deprecate(message?: string) {
  return (target: Function | Object, property?: string | symbol, descriptor?: unknown): void => {
    const isProtoOrStatic = typeof target === 'function';
    const className = isProtoOrStatic ? (target as Function).name : target.constructor.name;
    const accessSymbol = isProtoOrStatic ? `.${String(property)}` : `#${String(property)}`;

    // Class
    if (isClass(target, property, descriptor)) {
      console.debug(message || `Class \`${className}\` has been deprecated.`);

      // Method
    } else if (isMethod(target, property, descriptor)) {
      console.debug(message || `Method \`${className + accessSymbol}()\` has been deprecated.`);

      // Property
    } else if (isProperty(target, property, descriptor)) {
      console.debug(message || `Property \`${className + accessSymbol}\` has been deprecated.`);

      // Param (Babel/Jest doesnt support them)
    } /* istanbul ignore next */ else if (isParam(target, property, descriptor)) {
      console.debug(
        message ||
          `Parameter ${descriptor} for \`${className + accessSymbol}()\` has been deprecated.`,
      );
    }
  };
}
