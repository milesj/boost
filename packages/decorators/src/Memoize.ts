/* eslint-disable babel/no-invalid-this */

import isMethod from './helpers/isMethod';

export type MemoizedFunction<T> = (...args: unknown[]) => T;

export type MemoizeHasher = (...args: unknown[]) => string;

export type MemoizeCache<T> = Map<
  string,
  {
    time?: number | null;
    value: T;
  }
>;

export interface MemoizeOptions<T> {
  cache?: MemoizeCache<T> | null;
  expires?: number;
  hasher?: MemoizeHasher;
}

function hasher(...args: unknown[]): string {
  return JSON.stringify(args);
}

function createMemoizer<T>(
  method: MemoizedFunction<T>,
  rootCache: WeakMap<Function, MemoizeCache<T>>,
  options: Required<MemoizeOptions<T>>,
): MemoizedFunction<T> {
  // Must be a normal function as we require `this`
  return function memoizer(this: Function, ...args: unknown[]) {
    // Extract the cache for this specific instance
    let cache = rootCache.get(this);

    if (!cache) {
      cache = options.cache ? new Map(options.cache) : new Map();
      rootCache.set(this, cache);
    }

    // Hash the key and check the cache
    const key = options.hasher(...args);
    const item = cache.get(key);

    if (item && (!item.time || (typeof item.time === 'number' && item.time > Date.now()))) {
      return item.value;
    }

    // No cache so execute and cache the result
    const value = method.apply(this, args);
    const time = options.expires > 0 ? Date.now() + options.expires : null;

    cache.set(key, {
      time,
      value,
    });

    // Only cache if successful
    if (value instanceof Promise) {
      value.catch(() => cache?.delete(key));
    }

    return value;
  };
}

export default function Memoize<T>(
  options: MemoizeHasher | MemoizeOptions<T> = {},
): MethodDecorator {
  // eslint-disable-next-line complexity
  return (target, property, descriptor) => {
    if (__DEV__) {
      if (
        !isMethod(target, property, descriptor) ||
        (!('value' in descriptor && typeof descriptor.value === 'function') &&
          !('get' in descriptor && typeof descriptor.get === 'function'))
      ) {
        throw new TypeError(`\`@Memoize\` may only be applied to class methods or getters.`);
      }
    }

    const config = {
      cache: null,
      expires: 0,
      hasher,
      ...(typeof options === 'function' ? { hasher: options } : options),
    };

    if (__DEV__) {
      if (config.cache && !(config.cache instanceof Map)) {
        throw new Error('`cache` must be an instance of `Map`.');
      }

      if (typeof config.expires !== 'number' || config.expires < 0) {
        throw new Error('`expires` must be a number greater than or equal to 0.');
      }

      if (typeof config.hasher !== 'function') {
        throw new TypeError('`hasher` must be a function.');
      }
    }

    // We must use a map as all class instances would share the
    // same cache otherwise. Probability of collision is high.
    const rootCache = new WeakMap<Function, MemoizeCache<T>>();

    if (descriptor.get) {
      // @ts-expect-error
      descriptor.get = createMemoizer<T>(descriptor.get, rootCache, config);
    } else if (descriptor.value) {
      // @ts-expect-error
      descriptor.value = createMemoizer<T>(descriptor.value, rootCache, config);
    }
  };
}
