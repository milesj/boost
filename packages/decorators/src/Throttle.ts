import { isMethod } from './helpers/isMethod';

export function Throttle(delay: number): MethodDecorator {
	return (target, property, descriptor) => {
		if (
			__DEV__ &&
			(!isMethod(target, property, descriptor) ||
				!('value' in descriptor && typeof descriptor.value === 'function'))
		) {
			throw new TypeError(`\`@Throttle\` may only be applied to class methods.`);
		}

		// We must use a map as all class instances would share the
		// same boolean value otherwise.
		const throttling = new WeakMap<Function, boolean>();

		// Overwrite the value function with a new throttled function
		const func = descriptor.value;

		// @ts-expect-error Override generic
		descriptor.value = function throttle(this: Function, ...args: unknown[]) {
			if (throttling.get(this)) {
				return;
			}

			(func as unknown as Function).apply(this, args);
			throttling.set(this, true);

			setTimeout(() => {
				throttling.delete(this);
			}, delay);
		};
	};
}
