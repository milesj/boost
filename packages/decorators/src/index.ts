/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export { default as Bind } from './Bind';
export { default as Debounce } from './Debounce';
export { default as Deprecate } from './Deprecate';
export { default as Memoize } from './Memoize';
export { default as Throttle } from './Throttle';
