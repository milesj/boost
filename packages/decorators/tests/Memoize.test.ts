import { Memoize } from '../src';
import { vi, describe, it, expect, Mock } from 'vitest';

function sleep(time: number): Promise<void> {
	return new Promise((resolve) => {
		setTimeout(resolve, time);
	});
}

describe('@Memoize()', () => {
	// Date.now() is a bit too deterministic, as tests run so fast
	// and use the same timestamp. We use a counter to increase uniqueness.
	let count = 0;

	class Test {
		spy: Mock;

		constructor(spy: Mock) {
			this.spy = spy;
		}

		@Memoize()
		get getter(): number {
			this.spy();

			return Date.now() + this.inc();
		}

		@Memoize()
		noArgs(): number {
			this.spy();

			return Date.now() + this.inc();
		}

		@Memoize()
		oneArg(a: string): string {
			this.spy();

			return a.toUpperCase();
		}

		@Memoize()
		manyArgs(a: string, b: number, c: boolean): string {
			this.spy();

			// eslint-disable-next-line @typescript-eslint/restrict-plus-operands
			return a + b + c;
		}

		@Memoize()
		restArgs(...args: unknown[]): number {
			this.spy();

			return Date.now() + this.inc();
		}

		inc(): number {
			count += 1;

			return count;
		}
	}

	it('errors if applied to a class', () => {
		expect(() => {
			// @ts-expect-error Allow decorator here
			@Memoize()
			class TestClass {}

			return TestClass;
		}).toThrowErrorMatchingSnapshot();
	});

	it('errors if applied to a property', () => {
		expect(
			() =>
				class TestProp {
					// @ts-expect-error Allow decorator here
					@Memoize()
					value = 123;
				},
		).toThrowErrorMatchingSnapshot();
	});

	it('errors if `cache` is not a map', () => {
		expect(() => {
			class TestClass {
				// @ts-expect-error Allow decorator here
				@Memoize({ cache: {} })
				test() {}
			}

			return TestClass;
		}).toThrowErrorMatchingSnapshot();
	});

	it('errors if `expires` is not a number', () => {
		expect(() => {
			class TestClass {
				// @ts-expect-error Invalid type
				@Memoize({ expires: 'abc' })
				test() {}
			}

			return TestClass;
		}).toThrowErrorMatchingSnapshot();
	});

	it('errors if `expires` is negative', () => {
		expect(() => {
			class TestClass {
				@Memoize({ expires: -123 })
				test() {}
			}

			return TestClass;
		}).toThrowErrorMatchingSnapshot();
	});

	it('errors if `hasher` is not a function', () => {
		expect(() => {
			class TestClass {
				// @ts-expect-error Invalid type
				@Memoize({ hasher: 123 })
				test() {}
			}

			return TestClass;
		}).toThrowErrorMatchingSnapshot();
	});

	it('doesnt leak to separate instances', () => {
		const spy1 = vi.fn();
		const spy2 = vi.fn();

		const a = new Test(spy1);
		const b = new Test(spy2);

		a.noArgs();
		b.noArgs();

		expect(spy1).toHaveBeenCalledTimes(1);
		expect(spy2).toHaveBeenCalledTimes(1);
		expect(a).not.toBe(b);
	});

	it('caches result for getters', () => {
		const spy = vi.fn();
		const test = new Test(spy);

		const a = test.getter;
		const b = test.getter;
		const c = test.getter;

		expect(spy).toHaveBeenCalledTimes(1);
		expect(b).toBe(a);
		expect(c).toBe(a);
	});

	it('caches result for 0 arguments', () => {
		const spy = vi.fn();
		const test = new Test(spy);

		const a = test.noArgs();
		const b = test.noArgs();
		const c = test.noArgs();

		expect(spy).toHaveBeenCalledTimes(1);
		expect(b).toBe(a);
		expect(c).toBe(a);
	});

	it('caches result for 1 argument', () => {
		const spy = vi.fn();
		const test = new Test(spy);

		const a = test.oneArg('abc');
		const b = test.oneArg('abc');
		const c = test.oneArg('abc');

		expect(spy).toHaveBeenCalledTimes(1);
		expect(b).toBe(a);
		expect(c).toBe(a);

		const d = test.oneArg('xyz');
		const e = test.oneArg('xyz');

		expect(spy).toHaveBeenCalledTimes(2);
		expect(d).not.toBe(a);
		expect(e).toBe(d);
	});

	it('caches result for many arguments', () => {
		const spy = vi.fn();
		const test = new Test(spy);

		const a = test.manyArgs('abc', 1, true);
		const b = test.manyArgs('abc', 1, true);
		const c = test.manyArgs('abc', 1, true);

		expect(spy).toHaveBeenCalledTimes(1);
		expect(b).toBe(a);
		expect(c).toBe(a);

		const d = test.manyArgs('abc', 1, false);
		const e = test.manyArgs('abc', 2, true);
		const f = test.manyArgs('xyz', 1, true);
		const g = test.manyArgs('abc', 1, true);

		expect(spy).toHaveBeenCalledTimes(4);
		expect(d).not.toBe(a);
		expect(e).not.toBe(a);
		expect(f).not.toBe(a);
		expect(g).toBe(a);
	});

	it('caches result for rest arguments', () => {
		const spy = vi.fn();
		const test = new Test(spy);

		const a = test.restArgs('abc', 1, true, {});
		const b = test.restArgs('abc', 1, true, {});
		const c = test.restArgs('abc', 1, true, {});

		expect(spy).toHaveBeenCalledTimes(1);
		expect(b).toBe(a);
		expect(c).toBe(a);

		const d = test.restArgs('abc');
		const e = test.restArgs('abc', 3);
		const f = test.restArgs('xyz', 1, true, {}, []);
		const g = test.restArgs('abc', 1, true, {});

		expect(spy).toHaveBeenCalledTimes(4);
		expect(d).not.toBe(a);
		expect(e).not.toBe(a);
		expect(f).not.toBe(a);
		expect(g).toBe(a);
	});

	it('returns cache for same object argument structure', () => {
		const spy = vi.fn();
		const test = new Test(spy);

		const a = test.restArgs({ foo: 123 });
		const b = test.restArgs({ foo: 123 });
		const c = test.restArgs({ foo: 123 });

		expect(spy).toHaveBeenCalledTimes(1);
		expect(b).toBe(a);
		expect(c).toBe(a);

		const d = test.restArgs({ foo: 456 });
		const e = test.restArgs({ foo: 456 });
		const f = test.restArgs({ bar: 'abc' });

		expect(spy).toHaveBeenCalledTimes(3);
		expect(d).not.toBe(a);
		expect(e).toBe(d);
		expect(f).not.toBe(a);
		expect(f).not.toBe(e);
	});

	it('returns cache for same array argument structure', () => {
		const spy = vi.fn();
		const test = new Test(spy);

		const a = test.restArgs([123]);
		const b = test.restArgs([123]);
		const c = test.restArgs([123]);

		expect(spy).toHaveBeenCalledTimes(1);
		expect(b).toBe(a);
		expect(c).toBe(a);

		const d = test.restArgs([456]);
		const e = test.restArgs([456]);
		const f = test.restArgs(['abc']);

		expect(spy).toHaveBeenCalledTimes(3);
		expect(d).not.toBe(a);
		expect(e).toBe(d);
		expect(f).not.toBe(a);
		expect(f).not.toBe(e);
	});

	describe('cache map', () => {
		it('can provide a custom cache map', () => {
			const spy = vi.fn();
			const cache = new Map();
			cache.set('[]', { value: 'pre-cached value' });

			class TestCache {
				@Memoize({ cache })
				method() {
					spy();

					return 'value';
				}
			}

			const test = new TestCache();

			const a = test.method();
			const b = test.method();
			const c = test.method();

			expect(spy).toHaveBeenCalledTimes(0);
			expect(a).toBe('pre-cached value');
			expect(b).toBe(a);
			expect(c).toBe(a);
		});
	});

	describe('hash function', () => {
		function hasher() {
			return 'always same key';
		}

		it('can provide a custom hash function', () => {
			const spy = vi.fn();

			class TestHash {
				@Memoize({ hasher })
				method(...args: unknown[]) {
					spy();

					return Date.now();
				}
			}

			const test = new TestHash();

			// Differents args (cache key), but same results
			const a = test.method(1, 2, 3);
			const b = test.method('a', 'b', 'c');
			const c = test.method(1, 'a');

			expect(spy).toHaveBeenCalledTimes(1);
			expect(b).toBe(a);
			expect(c).toBe(a);
		});

		it('can provide a custom hash function by passing it directly to the decorator', () => {
			const spy = vi.fn();

			class TestHash {
				@Memoize(hasher)
				method(...args: unknown[]) {
					spy();

					return Date.now();
				}
			}

			const test = new TestHash();

			// Differents args (cache key), but same results
			const a = test.method(1, 2, 3);
			const b = test.method('a', 'b', 'c');
			const c = test.method(1, 'a');

			expect(spy).toHaveBeenCalledTimes(1);
			expect(b).toBe(a);
			expect(c).toBe(a);
		});
	});

	describe('expirations', () => {
		it('will bypass cache when an expiration time has passed', async () => {
			const spy = vi.fn();

			class TestExpires {
				@Memoize({ expires: 100 })
				method() {
					spy();

					return Date.now();
				}
			}

			const test = new TestExpires();

			const a = test.method();
			const b = test.method();
			const c = test.method();

			expect(spy).toHaveBeenCalledTimes(1);
			expect(b).toBe(a);
			expect(c).toBe(a);

			await sleep(110);

			const d = test.method();
			const e = test.method();
			const f = test.method();

			expect(spy).toHaveBeenCalledTimes(2);
			expect(d).not.toBe(a);
			expect(e).toBe(d);
			expect(f).toBe(d);
		});
	});

	describe('async/promises', () => {
		class TestAsync extends Test {
			@Memoize()
			async resolvedPromise() {
				this.spy();

				const value = await Promise.resolve(Date.now());

				return value;
			}

			@Memoize()
			async rejectedPromise(): Promise<unknown> {
				this.spy();

				await Promise.resolve(Date.now());

				throw new Error('Failed!');
			}
		}

		it('caches and reuses the same promise', async () => {
			const spy = vi.fn();
			const test = new TestAsync(spy);

			const a = test.resolvedPromise();
			const b = test.resolvedPromise();
			const c = test.resolvedPromise();

			expect(spy).toHaveBeenCalledTimes(1);
			expect(b).toBe(a);
			expect(c).toBe(a);

			const result = await a;

			await expect(b).resolves.toBe(result);
			await expect(c).resolves.toBe(result);

			const another = await test.resolvedPromise();

			expect(spy).toHaveBeenCalledTimes(1);
			expect(another).toBe(result);
		});

		it('deletes the cache if promise is rejected', async () => {
			expect.assertions(4);

			const spy = vi.fn();
			const test = new TestAsync(spy);

			const a = test.rejectedPromise();
			const b = test.rejectedPromise();
			const c = test.rejectedPromise();

			expect(spy).toHaveBeenCalledTimes(1);
			expect(b).toBe(a);
			expect(c).toBe(a);

			try {
				const result = await a;

				// Should not run
				await expect(b).resolves.toBe(result);
				await expect(c).resolves.toBe(result);
			} catch {
				// Ignore
			}

			try {
				await test.rejectedPromise();
			} catch {
				expect(spy).toHaveBeenCalledTimes(2);
			}
		});
	});
});
