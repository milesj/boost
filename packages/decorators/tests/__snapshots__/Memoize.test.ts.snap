// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Memoize() errors if \`cache\` is not a map 1`] = `"\`cache\` must be an instance of \`Map\`."`;

exports[`@Memoize() errors if \`expires\` is negative 1`] = `"\`expires\` must be a number greater than or equal to 0."`;

exports[`@Memoize() errors if \`expires\` is not a number 1`] = `"\`expires\` must be a number greater than or equal to 0."`;

exports[`@Memoize() errors if \`hasher\` is not a function 1`] = `"\`hasher\` must be a function."`;

exports[`@Memoize() errors if applied to a class 1`] = `"\`@Memoize\` may only be applied to class methods or getters."`;

exports[`@Memoize() errors if applied to a property 1`] = `"\`@Memoize\` may only be applied to class methods or getters."`;
